test/functorch/test_aotdispatch.py

```diff
@@ -1030,10 +1030,13 @@ def f(a, b):
             torch.ones(8, 8, device='cuda', requires_grad=True),
             torch.ones(1, 4, 1, device='cuda', requires_grad=True),
         ]
+        import gc
+        gc.collect()
         mem_before = torch.cuda.memory_allocated()
         f_compiled(*inps)
+        gc.collect()
```

Comment:
cc @albanD (and maybe @colesbury). I put all of the current context of this PR in the description. It appears to fix the ref cycle test that I added in `test_autograd.py`, which is great.

However, it fails this test in two interesting ways:

(1) When I update this test just a bit, it segfaults. I looked at the stacktrace, and it's inside a call to `tp_clear()`. It's calling the destructor on a `SavedVariable` that has a `DifferentiableViewMeta`. That then calls the destructor on the view's `.base_`, which is where the segfault happens (presumably the base was already freed?). I don't have the stack trace anymore but I can try to get it again if that's helpful.

The "change" I made to induce the segfault was just this:
```
...
mem_before = torch.cuda.memory_allocated()
f_compiled(*inps)
f_compiled(*inps)
f_compiled(*inps)
f_compiled(*inps)
gc.collect()
...
```

(2) When I run this test as-is, it appears to free input #2's memory.

I see `mem_before = 1024`. That's a bit strange to me, but it looks like we allocate 512B for each input tensor (maybe that has to do with the cuda caching allocator having a minimum block size). `mem_after` however has a value of `512`, which is pretty suspicious. I tried printing `input[0]` which worked, and printing `input[1]` segfaults - so I'm probably accidentally convincing the python gc that `input[1]` is part of this cycle.

Reply:
Update: judging from CI, my changes are causing many segfaults due to freeing tensors too aggressively.

Reply:
I also confirmed that in the repro test, `input[1]` was saved_for_backward by the autograd.Function. The final AOTAutograd graph looks like this:
```
def forward(self, primals_1, primals_2):
    add = torch.ops.aten.add.Tensor(primals_1, primals_1);  primals_1 = None
    split_with_sizes = torch.ops.aten.split_with_sizes.default(add, [4, 4], 1)
    getitem_1 = split_with_sizes[1];  split_with_sizes = None
    unsqueeze = torch.ops.aten.unsqueeze.default(getitem_1, -1)
    mul = torch.ops.aten.mul.Tensor(unsqueeze, primals_2)
    return [getitem_1, mul, add, primals_2, unsqueeze]
```

Where out of the 5 outputs, the last 2 are saved for backward